### PR TITLE
[Merged by Bors] - chore: split `Order`ed instances for subobjects into separate files 

### PR DESCRIPTION
**Pull request by mattrobball**
Moving these to separate files should make typeclass synthesis less expensive. Additionally two of them are quite long and this shrinks them slightly.

This handles:
* `Submonoid`
* `Subgroup`
* `Subsemiring`
* `Subring`
* `Subfield`
* `Submodule`
* `Subalgebra`

This also moves `Units.posSubgroup` into its own file.

The copyright headers are from:
* https://github.com/leanprover-community/mathlib/pull/6489
* https://github.com/leanprover-community/mathlib/pull/8466

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
!bench

**Review by eric-wieser — APPROVED** [review]
I think the copyright / authors should probably come from https://github.com/leanprover-community/mathlib/pull/6489 for all these new files (2021 and @adomani).

Is there a similar change to be made for submonoids?

File: Mathlib/RingTheory/Subsemiring/Order.lean [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+/-!
+
+`Order`ed instances for `SubsemiringClass` and `Subsemiring`.
+
+-/
+
+namespace SubsemiringClass
+
+-- variable {R : Type u} {S : Type v} {T : Type w} [NonAssocSemiring R] (M : Submonoid R)
+-- variable [SetLike S R] [hSR : SubsemiringClass S R] (s : S)
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Subsemiring/Order.lean:19]: Dead code?

**Comment by eric-wieser** [comment]
@YaelDillies  noted on Zulip that they want to take a look

**Comment by mattrobball** [comment]
@eric-wieser feel free to push your suggested changes 

There are no ordered instances in `Submonoid` from what I remember

File: Mathlib/RingTheory/Subring/Order.lean [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#align subring_class.to_linear_ordered_comm_ring SubringClass.toLinearOrderedCommRing
+
+end SubringClass
+
+namespace Subring
+
+variable {R : Type*}
+
+/-- A subring of an `OrderedRing` is an `OrderedRing`. -/
+instance toOrderedRing  [OrderedRing R] (s : Subring R) : OrderedRing s :=
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Subring/Order.lean:60]:
```suggestion
instance toOrderedRing [OrderedRing R] (s : Subring R) : OrderedRing s :=
```
and below

**Review by YaelDillies — COMMENTED** [review]
Please update the copyright headers as indicated. The rest looks good.

**Comment by eric-wieser** [comment]
> There are no ordered instances in `Submonoid` from what I remember

Yes there are! I just pushed changes that extracted those too.

File: Mathlib/Algebra/Module/Submodule/Order.lean [hunk 3]
```diff
@@ -0,0 +1,68 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.GroupTheory.Subgroup.Order
+import Mathlib.GroupTheory.Submonoid.Order
```
Comment (eric-wieser) [inline Mathlib/Algebra/Module/Submodule/Order.lean:8]: I deliberately added these imports even though they're possibly redundant, to reduce confusion about mathlib putting orders on stronger structures without weaker structures.

I think we should tell `shake` to ignore these if it complains.

**Comment by eric-wieser** [comment]
> Oops! Shows what good my memory is. Am I correct that still the only files downstream of these are themselves? 

CI will tell I think.

Should we split `Subfield` too, or is that far enough along the hierarchy that we don't care?

**Comment by mattrobball** [comment]
Yeah. I think we should.

**Comment by mattrobball** [comment]
`Units.posSubgroup` should be moved out of there.

**Comment by mattrobball** [comment]
!bench

**Comment by eric-wieser** [comment]
I don't understand why we have a benchmark result if the build is failing

**Comment by eric-wieser** [comment]
Ah nevermind, github was showing me stale info; I'll run `shake update`

**Comment by eric-wieser** [comment]
!bench

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
Ok this actually bump the top line number to -350B (from 275B). 

@eric-wieser if you happy, please send it off

File: Mathlib/RingTheory/Polynomial/Cyclotomic/Expand.lean [hunk 4]
```diff
@@ -8,6 +8,7 @@ import Mathlib.Data.ZMod.Algebra
 
 #align_import ring_theory.polynomial.cyclotomic.expand from "leanprover-community/mathlib"@"0723536a0522d24fc2f159a096fb3304bef77472"
 
+
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Polynomial/Cyclotomic/Expand.lean:11]:
```suggestion
```

File: Mathlib/RingTheory/Subring/Units.lean [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/-
+Copyright (c) 2021 Chris Birbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birbeck
```
Comment (eric-wieser) [inline Mathlib/RingTheory/Subring/Units.lean:4]: Can you mention where this came from in the PR description?

Comment (eric-wieser) [reply Mathlib/RingTheory/Subring/Units.lean:4]: I've done the archaeology myself, now in the description.

**Comment by eric-wieser** [comment]
bors merge

Thanks!

File: Mathlib/RingTheory/Subring/Units.lean (hunk 5 above)
Comment (kbuzzard) [reply Mathlib/RingTheory/Subring/Units.lean:4]: This is surely "Birkbeck"?

Comment (kbuzzard) [reply Mathlib/RingTheory/Subring/Units.lean:4]: Oops, too late.

Comment (kbuzzard) [reply Mathlib/RingTheory/Subring/Units.lean:4]: #10966

**Pull request closed without merging**